### ISSUE TITLE
Error specifying both base directory and a globbing pattern in an exclude directive in Windows

### ISSUE THREAD

**Issue by desseim** [rdiff-backup/rdiff-backup#620]
Not sure it's a bug, it may be usage problem, but why does this command work fine:
```
rdiff-backup.exe backup --exclude-symbolic-links --exclude '**/Users/*/AppData/Roaming/Mozilla/Firefox/' --include 'C:/Users/TheUser' --exclude 'C:/' C:/ D:/
```
while this similar one fails with the following error:
```
rdiff-backup.exe backup --exclude-symbolic-links --exclude 'C:/Users/*/AppData/Roaming/Mozilla/Firefox/' --include 'C:/Users/TheUser' --exclude 'C:/' C:/ D:/

Fatal Error: Fatal Error: The file specification
At line:1 char:1
+ rdiff-backup.exe backup --exclude-symbolic-links --exclude 'C:/Users/ ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (2021-08-27 22:1...e specification:String) [], RemoteException
    + FullyQualifiedErrorId : NativeCommandError
 
    'b'C:/Users/*/AppData/Roaming/Mozilla/Firefox/''
cannot match any files in the base directory
    'b'C:/''
Useful file specifications begin with the base directory or some
pattern (such as '**') which matches the base directory.
```
?

I tried executing the command from the context of the base directory (`cd C:\`), and using escaping backslashes as suggested in the Windows README but it didn't make a difference.

`rdiff-backup.exe backup --exclude-symbolic-links --exclude 'C:/Users/TheUser/AppData/Roaming/Mozilla/Firefox/' --include 'C:/Users/TheUser' --exclude 'C:/' C:/ D:/` works fine as well so it seems to be a problem when the base directory is specified in a pattern containing globbing characters.

Is this syntax supposed not to be acceptable?

-----

Environment:
```yaml
---
exec:
  api_version:
    actual: 0
    default: 200
    max: 201
    min: 200
  argv:
  - C:\Program Files\rdiff-backup\rdiff-backup.exe
  - info
  version: 2.1.0a1
python:
  executable: C:\Program Files\rdiff-backup\rdiff-backup.exe
  name: cpython
  version: 3.9.1
system:
  fs_encoding: utf-8
  platform: Windows-10-10.0.19041-SP0
...
```

**Comment by viertelb** [comment]
This probably has something to do with how the quotes are interpreted by the terminal. 

I can use this line in a linux bash script with no problems:

```
sudo runuser -u ben -- rdiff-backup -v 9 --include "**/...*" --exclude "**/*DS_store" --exclude "**/.*.*" $TRG/$1 &>> $FILE_LOG;
```

where `$1` is an argument for the specific directory I want to backup. I give it when running the script. `$TRG` is the underlying path and `$FILE_LOG` is the logfile. 

I now tried to put the command without the paths in a variable in the bash script: 

```
CMD='sudo runuser -u ben -- rdiff-backup -v 9 --include "**/...*" --exclude "**/*DS_store" --exclude "**/.*.*"'
```

Then I run it in the script with 

```
$CMD $SRC/$1 $TRG/$1 &>> $FILE_LOG;
```

When I do this I get: 

```
Fatal Error: Fatal Error: The file specification 'b"'**"' cannot match any files in the 
base directory 'b'/mnt/mybook/test/AUDIO/Musik''
```

When I remove the `include` and `exclude`I get no errors and it will backup a file that was excluded before: `Regular copying (b'._.DS_Store',) to /mnt/HGST/RDIFFREMOTE/Musik/rdiff-backup.tmp.1`

**Comment by viertelb** [comment]
wenn I only use on `--exclude "test*"` I get

```
Fatal Error: Fatal Error: The file specification
    'b'"test*"''
cannot match any files in the base directory
    'b'/mnt/mybook/test/AUDIO/Musik''
Useful file specifications begin with the base directory or some
pattern (such as '**') which matches the base directory.

```

**Comment by viertelb** [comment]
omiting the quotes (`--exclude test*`) also throws an error:

```
Fatal Error: Fatal Error: The file specification
    'b'test*''
cannot match any files in the base directory
    'b'/mnt/mybook/test/AUDIO/Musik''
Useful file specifications begin with the base directory or some
pattern (such as '**') which matches the base directory.
```

**Comment by desseim** [comment]
@viertelb

The original issue is unrelated to shell quoting, as exclude directives, one which "works" and one which doesn't, *both* include globbing patterns and are *both* quoted in the same manner. The only difference is whether the base directory (the drive letter under Windows) is explicitly specified, or implied with a globbing pattern. Quotes remain the same, globbing patterns in other parts of the path too. That is, `--exclude '**/Users/*/AppData/Roaming/Mozilla/Firefox/'` "works" while `--exclude 'C:/Users/*/AppData/Roaming/Mozilla/Firefox/'` doesn't.

There are a bunch of shell shenanigans implicated in the several examples you gave ([quoting](<https://mywiki.wooledge.org/Quotes>), [word splitting](<https://mywiki.wooledge.org/Arguments#Word_Splitting>), [putting a command in a variable](<https://mywiki.wooledge.org/BashFAQ/050>)...). Generally, you'd want to quote arguments to include/exclude directives, or the shell will (try to) expand them into potentially multiple arguments (only the first of which rdiff-backup will interpret as an argument to the directive...).

**Comment by desseim** [comment]
A slightly different example, but which I suspect has the same root cause:

```console
> & 'C:\Program Files\rdiff-backup-2.1.0a1\rdiff-backup.exe' backup --exclude '**/Users/TheUser/test/src/dir\ 1/file1.txt' --include 'C:/Users/TheUser/test/src' --exclude 'C:/' C:/ D:/
Symbolic links excluded by default on Windows
> # backup successful

> & 'C:\Program Files\rdiff-backup-2.1.0a1\rdiff-backup.exe' backup --exclude 'C:/Users/TheUser/test/src/dir\ 1/file1.txt' --include 'C:/Users/TheUser/test/src' --exclude 'C:/' C:/ D:/
Symbolic links excluded by default on Windows
Fatal Error: Fatal Error: The file specification
    'b'C:/Users/TheUser/test/src/dir\\ 1/file1.txt''
cannot match any files in the base directory
    'b'C:/''
Useful file specifications begin with the base directory or some
pattern (such as '**') which matches the base directory.
> # backup failed

> & 'C:\Program Files\rdiff-backup-2.1.0a1\rdiff-backup.exe' backup --exclude 'C:/Users/TheUser/test/src/dir 1/file1.txt' --include 'C:/Users/TheUser/test/src' --exclude 'C:/' C:/ D:/
Symbolic links excluded by default on Windows
> # backup successful
```

**Comment by ericzolf** [comment]
I think it's been fixed a long time ago but I need to check with the latest code.

**Comment by ericzolf** [comment]
@desseim can you please try again the commands _without_ single quotes, I think they're the issue. Either don't use them, or use double quotes. The Windows shell doesn't handle single quotes, and then they become part of the path for rdiff-backup so of course they don't match. None of your examples work if I use single quotes, they all work if I don't use any quote or _double_quotes.

I can only guess that you've done the same with your working examples, but did copy&paste incorrectly. If you still think that's an issue, perhaps you can tell me which shell you're actually using under Windows. I'm using the standard cmd.

**Comment by desseim** [comment]
Thanks for having a look @ericzolf.

There is no copy-pasting error, so quoting is not the issue (the 1st and 3rd example of my last post _do_ succeed); I'm just running them with PowerShell. Here are the results of the same examples run with `CMD`:

```console
> "C:\Program Files\rdiff-backup-2.1.0a1\rdiff-backup.exe" backup --exclude "**/Users/TheUser/test/src/dir\ 1/file1.txt" --include "C:/Users/TheUser/test/src" --exclude "C:/" C:/ D:/
Symbolic links excluded by default on Windows
> # backup successful

> "C:\Program Files\rdiff-backup-2.1.0a1\rdiff-backup.exe" backup --exclude "C:/Users/TheUser/test/src/dir\ 1/file1.txt" --include "C:/Users/TheUser/test/src" --exclude "C:/" C:/ D:/
Symbolic links excluded by default on Windows
Fatal Error: Fatal Error: The file specification
    'b'C:/Users/TheUser/test/src/dir\\ 1/file1.txt''
cannot match any files in the base directory
    'b'C:/''
Useful file specifications begin with the base directory or some
pattern (such as '**') which matches the base directory.
> # backup failed

> "C:\Program Files\rdiff-backup-2.1.0a1\rdiff-backup.exe" backup --exclude "C:/Users/TheUser/test/src/dir 1/file1.txt" --include "C:/Users/TheUser/test/src" --exclude "C:/" C:/ D:/
Symbolic links excluded by default on Windows
> # backup successful
```

Also, the issue reproduces with include / exclude directives specified in a file (`--exclude-filelist`); I only translated them to `--exclude` / `--include` options for bug reporting.

Maybe you haven't reproduced the same directory structure for your tests? Note the escaped space in the exclude directive's file path.

I see two potentially problematic inconsistencies in these examples:

1. `--exclude "C:/path\ with\ escaped\ spaces"` fails under Windows while it works as expected under Linux.
2. Under Windows, `--exclude "C:/path\ with\ escaped\ spaces"` fails while `--exclude "**/path\ with\ escaped\ spaces"` works as expected.

Hope this clarifies it a bit.

Please note that the original issue is a bit different as it pertains to globbing characters in the path instead of escaping characters, but I suspect the root cause might be the same.

**Comment by ericzolf** [comment]
Thanks for answering so fast. I won't be so fast so if you find the time to test again with the latest alpha version 2.1.1a0, it would be much appreciated as I already fixed similar issues. Else, I plan to release a beta version soonish, so we can wait until then to pursue the issue.

**Comment by desseim** [comment]
No problem, I appreciate the following up.

Here are the output for the same commands run with `v2.1.1a0` in `CMD`:

```console
> "C:\Program Files\rdiff-backup-2.1.1a0\rdiff-backup.exe" backup --exclude "**/Users/TheUser/test/src/dir\ 1/file1.txt" --include "C:/Users/TheUser/test/src" --exclude "C:/" C:/ D:/
NOTE: Symbolic links excluded by default on Windows
NOTE: Starting mirror from source path C:/ to destination path D:/
> # backup successful

> "C:\Program Files\rdiff-backup-2.1.1a0\rdiff-backup.exe" backup --exclude "C:/Users/TheUser/test/src/dir\ 1/file1.txt" --include "C:/Users/TheUser/test/src" --exclude "C:/" C:/ D:/
NOTE: Symbolic links excluded by default on Windows
Fatal Error: The file specification 'b'C:/Users/TheUser/test/src/dir\\ 1/file1.txt''
cannot match any files in the base directory 'b'C:/''.
Useful file specifications begin with the base directory or some
pattern (such as '**') which matches the base directory
> # backup failed

> "C:\Program Files\rdiff-backup-2.1.1a0\rdiff-backup.exe" backup --exclude "C:/Users/TheUser/test/src/dir 1/file1.txt" --include "C:/Users/TheUser/test/src" --exclude "C:/" C:/ D:/
NOTE: Symbolic links excluded by default on Windows
NOTE: Starting increment operation from source path C:/ to destination path D:/
> # backup successful
```

There were a few `ListError`s for `C:/$Recycle.bin` sub-directories which I have omitted as they don't seem relevant (they occurred with other versions too, it's not peculiar to `v2.1.1a0`).

**Comment by ericzolf** [comment]
OK, I'll try again during the upcoming beta phase. Thanks for confirming that the issue is still there, and (perhaps) identifying the underlying issue (blank in name?!).

**Comment by desseim** [comment]
FYI, I tried again with `v2.1.2b2` and obtained exactly the same result and output as [with `v2.1.1a0`](https://github.com/rdiff-backup/rdiff-backup/issues/620#issuecomment-1228814568).

**Comment by ericzolf** [comment]
Then it's on me. Thanks for taking the pain to verify.

**Comment by ericzolf** [comment]
Ah, wait, I think, I have been blind the whole time... Does the first example actually really exclude the file? I expect not, can you check? If you could also check the 3rd example, it would help.

Here my thinking:
* the 2nd example fails because it has quotes _and_ backslash and hence the backslash becomes part of the path, why it shouldn't match the base path is still unclear to me so there is definitely something weird.
* the 1st example doesn't fail the validation because of the double asterisk (such glob isn't checked, it could match anywhere) but the backslash should still be there and _not_ match `dir 1`.

In other words, the first two calls should rightly not match `dir 1`, but the 2nd one should also silently fail because the logic is to check that the exclude _can_ match a file, not that it does it, and there could be a file `dir\\ 1` theoretically (possibly translated to `dir/ 1` BTW...).

**Comment by ericzolf** [comment]
OK, I have proven myself wrong, this works (with a path below `C:`:

```
> rdiff-backup -v5 backup --exclude "C:/temp/from/src/dir\ 1/file1.txt" --include "C:/temp/from/src" --exclude C:/temp/from C:/temp/from C:/temp/bak2
[...]
NOTE: Starting mirror from source path C:/temp/from to destination path C:/temp/
bak2
* Processing file .
* Processing file src
* Processing file src/dir 1
* Processing file src/dir 1/file2.txt
* Processing file src/file3.txt
* Cleaning up
```

But this doesn't (using only `C:/`):

```
>rdiff-backup -v5 backup --exclude "C:/temp/from/src/dir\ 1/file1.txt" --include "C:/temp/from/src" --exclude C:/ C:/ C:/temp/bak12
[...]
Fatal Error: The file specification 'b'C:/temp/from/src/dir\\ 1/file1.txt''     
cannot match any files in the base directory 'b'C:/''.
Useful file specifications begin with the base directory or some
pattern (such as '**') which matches the base directory
* Cleaning up
```

And `rdiff-backup -v5 backup --exclude "**/temp/from/src/dir\ 1/file1.txt" --include "C:/temp/from/src" --exclude C:/ C:/ C:/temp/bak11` also does work correctly, excluding `file1` as expected (not by me, but other story). As well, `rdiff-backup -v5 backup --exclude "C:/temp/from/src/dir 1/file1.txt" --include "C:/temp/from/src" --exclude C:/ C:/ C:/temp/bak13` does work as expected.

So, for the bug to appear, there needs to be the backslash _and_ a source being the root of a drive `C:\`. Now I'm really wondering, but it's definitely on me.

**Comment by ericzolf** [comment]
I need to stop debugging, but for my notes: `if not include_sel_func(self.rpath): raise FilePrefixError(glob_str)` in `src/rdiff_backup/selection.py` possibly raises a a FilePrefixError (catched by `_parse_catch_error` and handled as such), even though it could be simply a glob/regex parsing error (due to the backslash). I'm still not sure it is the full story but it goes in the direction.

**Comment by ericzolf** [comment]
Note to myself (Windows shell history is a joke):

```
python -m pdb build\scripts-3.10\rdiff-backup -v5 backup --exclude "C:/temp/from/src/dir\ 1/file1.txt" --include "C:/temp/from/src" --exclude C:/ C:/ C:/temp/bak12
(Pdb) b c:\users\vagrant\develop\rdiff-backup\build\lib.win-amd64-cpython-310\rdiff_backup\selection.py:681 
Breakpoint 1 at c:\users\vagrant\develop\rdiff-backup\build\lib.win-amd64-cpython-310\rdiff_backup\selection.py:681
(Pdb) b c:\users\vagrant\develop\rdiff-backup\build\lib.win-amd64-cpython-310\rdiff_backup\selection.py:273 
Breakpoint 2 at c:\users\vagrant\develop\rdiff-backup\build\lib.win-amd64-cpython-310\rdiff_backup\selection.py:273
```

OK, got it I think, `_glob_get_prefix_res` should contain something similar for `X:`:
```
        # we must make exception for root "/", only dir to end in slash
        if prefixes[0] == b"":
            prefixes[0] = b"/"
```

**Comment by desseim** [comment]
I have only read the diff from the PR, but the change looks very consistent with my understanding of the issue.

I'll also test with my "production" use cases once Windows artifacts that include this change are released, and report here only in the case it wouldn't be fully fixed.

Thanks a lot for the in-depth debugging and fix!
